### ISSUE TITLE
OBS Studio v. 25.0.0 breaks streaming (handshake: unexpected RTMP version: 80)

### ISSUE THREAD

**Issue by andreipopovici** [alfg/docker-nginx-rtmp#56]
Title says it all, connection drops right after being established.

```
2020/03/18 17:29:10 [info] 7#7: *90 client connected '172.17.0.1'

2020/03/18 17:29:11 [info] 7#7: *90 handshake: unexpected RTMP version: 80, client: 172.17.0.1, server: 0.0.0.0:1935

2020/03/18 17:29:11 [info] 7#7: *90 handshake: error parsing challenge, client: 172.17.0.1, server: 0.0.0.0:1935

2020/03/18 17:29:11 [info] 7#7: *90 disconnect, client: 172.17.0.1, server: 0.0.0.0:1935

2020/03/18 17:29:11 [info] 7#7: *90 deleteStream, client: 172.17.0.1, server: 0.0.0.0:1935
```

Downgrading to OBS Studio 24.0.3 (in the middle of a live broadcast!!) fixed the issue. There's nothing in the [25.0.0 release notes](https://github.com/obsproject/obs-studio/releases/tag/25.0.0) that jumped out at me as being a probable cause.

**Comment by alfg** [comment]
Wow, ok. I'll check into this tonight. Thanks for reporting this.

**Comment by andreipopovici** [comment]
I was reminded the hard way never to update crucial software right before live production 😅

**Comment by alfg** [comment]
Hi @andreipopovici,

I'm trying OBS `25.0.0 `now with the latest docker `nginx-rtmp `image and it seems to work OK. Can you try pulling the latest image? There's been some changes within the last week or so.

**Comment by andreipopovici** [comment]
@alfg Will try later today. I have to admit I didn't pull latest, since I rebuilt to enable SSL certs. Is there an easier way to toggle SSL certs (and even changes to `nginx.conf`) without having to rebuild each time?

**Comment by alfg** [comment]
You can mount your `nginx.conf` as a volume, that way it won't need to rebuild the Dockerfile each time you want to test something.

For example, in `docker-compose.yml`:
```
    volumes:
      - ./certs:/opt/certs
      - ./nginx.conf:/etc/nginx/nginx.conf
```

**Comment by andreipopovici** [comment]
Thanks for the volumes tip! Works great.

I still get the same error with 25.0, with the `latest` image. I didn't mention that it's OBS Studio for Windows, not macOS, if it helps.

Funny thing is, OBS 25.0 continues to stream (shows `LIVE` as status), but pulling the HLS stream throws `open() "/opt/data/hls/test.m3u8" failed (2: No such file or directory)`.

**Comment by alfg** [comment]
Yeah, I tried on OBS Windows too and it seemed OK. That looks like an error from a previous build. 

Can you paste the Docker image ID from `docker images`?

**Comment by andreipopovici** [comment]
I think we can close this until I get a chance to reproduce. I'll stick with 24.0.3 on Windows until I get more time to test. Tip: never upgrade a production system before a scheduled broadcast 😅

**Comment by andreipopovici** [comment]
I had a chance to test again and I see the same error on Windows and macOS with the latest versions of OBS for each (25.0.4 and 25.0.7 respectively).

If it helps, here's my `nginx.conf`.

```
user nobody;
daemon off;
worker_processes auto;
worker_rlimit_nofile 100000;
error_log /dev/stdout info;

events {
    worker_connections 1024;
}

rtmp {
    server {
        listen 1935;
        chunk_size 8192;
		max_message 4M;
		buflen 5s;
		sync 50ms;

        application stream {
            live on;

            exec ffmpeg -i rtmp://localhost:1935/stream/$name
              -c:a libfdk_aac -b:a 192k -c:v libx264 -b:v 4000k -f flv -g 30 -r 30 -s 1920x1080 -preset superfast -profile:v baseline rtmp://localhost:1935/hls/$name_1080p4192kbs
              -c:a libfdk_aac -b:a 128k -c:v libx264 -b:v 2500k -f flv -g 30 -r 30 -s 1280x720 -preset superfast -profile:v baseline rtmp://localhost:1935/hls/$name_720p2628kbs
              -c:a libfdk_aac -b:a 128k -c:v libx264 -b:v 1000k -f flv -g 30 -r 30 -s 854x480 -preset superfast -profile:v baseline rtmp://localhost:1935/hls/$name_480p1128kbs
              -c:a libfdk_aac -b:a 64k -c:v libx264 -b:v 200k -f flv -g 15 -r 15 -s 426x240 -preset superfast -profile:v baseline rtmp://localhost:1935/hls/$name_240p264kbs;
        }

        application hls {
            live on;
            hls on;
            hls_fragment_naming system;
            hls_fragment 5;
            hls_playlist_length 90m;
	    	hls_continuous on;
            hls_path /opt/data/hls;
            hls_nested on;
	    	hls_cleanup on;
	    	hls_type event;

	    	exec_publish_done bash -c 'tee -a /opt/data/hls/${name}/index.m3u8 <<< "#EXT-X-ENDLIST"';

            hls_variant _1080p4192kbs BANDWIDTH=4192000,RESOLUTION=1920x1080;
            hls_variant _720p2628kbs BANDWIDTH=2628000,RESOLUTION=1280x720;
            hls_variant _480p1128kbs BANDWIDTH=1128000,RESOLUTION=854x480;
            hls_variant _240p264kbs BANDWIDTH=264000,RESOLUTION=426x240;
        }

        application dash {
            live on;
            dash on;
            dash_path /opt/data/dash;
        }
    }
}

http {
    log_format json_combined escape=json
        '{'
            '"time_iso8601":"$time_iso8601",'
            '"time_local":"$time_local",'
            '"remote_addr":"$remote_addr",'
            '"remote_user":"$remote_user",'
            '"request":"$request",'
            '"token":"$arg_token",'
            '"args":"$args",'
            '"quality":"$request_filename",'
            '"status": "$status",'
            '"http_referrer":"$http_referer",'
            '"http_user_agent":"$http_user_agent"'
        '}';
    access_log /opt/data/logs/access.log json_combined;

    ssl_ciphers         HIGH:!aNULL:!MD5;
    ssl_protocols       TLSv1 TLSv1.1 TLSv1.2;
    ssl_session_cache   shared:SSL:10m;
    ssl_session_timeout 10m;

    server {
        listen 443 ssl;
        ssl_certificate     /etc/letsencrypt/live/<redacted>/fullchain.pem;
        ssl_certificate_key /etc/letsencrypt/live/<redacted>/privkey.pem;

        location /hls {
            types {
                application/vnd.apple.mpegurl m3u8;
                video/mp2t ts;
            }
            root /opt/data;
            add_header Cache-Control no-cache;
            include cors.conf;
            auth_request /auth;
            secure_token token=$arg_token;
			secure_token_types application/vnd.apple.mpegurl;
        }

        location /dash {
            # Serve DASH fragments
            root /opt/data;
            add_header Cache-Control no-cache;
            include cors.conf;
            auth_request /auth;
            secure_token token=$arg_token;
			secure_token_types application/dash+xml;
        }

        location /live {
          alias /opt/data/hls;
          types {
              application/vnd.apple.mpegurl m3u8;
              video/mp2t ts;
          }
          add_header Cache-Control no-cache;
          include cors.conf;
          auth_request /auth;
          secure_token token=$arg_token;
          secure_token_types application/vnd.apple.mpegurl;
        }
        
        resolver 1.1.1.1 1.0.0.1 8.8.8.8 8.8.4.4 208.67.222.222 208.67.220.220 valid=60s;
	    resolver_timeout 2s;
        ssl_protocols TLSv1.2 TLSv1.1;

        location = /auth {
			internal;
			set $query '';
			if ($request_uri ~* "[^\?]+\?(.*)$") {
				set $query $1;
			}
			set $dev_host <redacted>;
			proxy_pass $dev_host/auth/session?$query;
			proxy_pass_request_body off;
			proxy_set_header Content-Length "";
			proxy_set_header X-Original-URI $request_uri;
            proxy_ssl_protocols TLSv1.2 TLSv1.1;
            proxy_ssl_server_name on;
        }
    }
}
```

I did add `nginx-vod-module` and `nginx-secure-token-module` to the compiled nginx in my Dockerfile, but I don't see how those would interfere with incoming RTMP streams and their connection. The error seems to happen [here](https://github.com/arut/nginx-rtmp-module/blob/master/ngx_rtmp_handshake.c#L290) and I'm not savvy enough to debug why `if (*b->pos != '\x03')` is not true or how to determine the `*b->pos` of my current RTMP stream, either on Windows or macOS.

**Comment by alfg** [comment]
I am not able to reproduce this issue, but leaving the issue open in case anyone else has the same problem and can report.

**Comment by andreipopovici** [comment]
I figured out what was wrong, so I'll document this here in case someone finds this issue in the future.

I had copied the OBS profile and settings folders from the Windows version to a Mac. Since the Windows version contains an option to use "new" networking code, the macOS version simply closed the connection as soon as it was established, due to not recognizing the config option.

Manually removing that option from the `.ini` files (or simply not cross-transferring settings) fixed the issue, I can now broadcast using the latest versions on both OSes.
